Flaky `test_resharding_deferred.py::test_resharding_transfer_deferred_points[down]`

https://github.com/qdrant/qdrant/actions/runs/23863480095/job/69576104444

[consensus-test-logs.zip](https://github.com/user-attachments/files/26431886/consensus-test-logs.zip)

```
=================================== FAILURES ===================================
________________ test_resharding_transfer_deferred_points[down] ________________

tmp_path = PosixPath('/tmp/pytest-of-runner/pytest-1/test_resharding_transfer_defer1')
direction = 'down'

    @pytest.mark.parametrize("direction", ["up", "down"])
    def test_resharding_transfer_deferred_points(tmp_path: pathlib.Path, direction: Literal["up", "down"]):
        """Resharding must preserve deferred points.
    
        Deferred points live in appendable segments but have an internal offset
        beyond the indexing threshold, making them invisible to reads until the
        segment is optimized. The resharding_stream_records transfer uses
        DeferredBehavior::IncludeAll to read them. The optimizer runs concurrently
        (required for wait=true on the last migration batch) while the migration
        reads from the same segments via the forward proxy.
    
        For "down", the removed shard's data is permanently deleted — any deferred
        points not transferred during migration are irrecoverably lost.
    
        For "up", deferred points on source shards must be included when migrating
        the subset of points that belong to the new shard.
    
        Steps:
        1. Creates a collection with prevent_unoptimized + disabled optimizers
        2. Upserts enough points to create deferred points in every shard
        3. Verifies deferred points exist (only a fraction visible)
        4. Starts resharding while optimizers are still disabled
        5. Enables optimizers right before migration (required for wait=true)
        6. Migrates points between source and target shards
        7. Completes resharding (commit hash rings + finish)
        8. Verifies ALL points are visible after resharding
        """
        assert_project_root()
    
        peer_api_uris, _, _ = start_cluster(tmp_path, N_PEERS)
    
        peer_ids = []
        for uri in peer_api_uris:
            peer_ids.append(get_cluster_info(uri)["peer_id"])
    
        create_deferred_collection(peer_api_uris[0])
        wait_collection_exists_and_active_on_all_peers(
            collection_name=COLLECTION_NAME,
            peer_api_uris=peer_api_uris,
        )
    
        # Upsert enough points to push beyond the indexing threshold on each shard.
        # With 3 shards, ~500 points per shard, well beyond the ~100 point threshold.
        total_points = 1500
        upsert_points(peer_api_uris[0], start_id=1, count=total_points, wait=False)
        time.sleep(3)
    
        # Verify deferred points exist: only a fraction should be visible
        visible = scroll_all(peer_api_uris[0])
        visible_count = len(visible)
        assert visible_count > 0, "Some points should be visible (within threshold)"
        assert visible_count < total_points, (
            f"Not all points should be visible (most are deferred), "
            f"got {visible_count}/{total_points}"
        )
                json={
                    "finish_migrating_points": {
                        "peer_id": target_peer_id,
                        "shard_id": target_shard_id,
                    }
                },
            )
            assert_http_ok(resp)
            time.sleep(1)
    
        # Complete resharding: commit hash rings and finish.
        resp = commit_read_hashring(peer_api_uris[0])
        assert_http_ok(resp)
    
        if direction == "up":
            # Clean up source shards between read and write hash ring commits:
            # delete points that were migrated to the new shard and no longer
            # belong on the source under the new hash ring.
            for shard_id in range(target_shard_id):
                peer_id, peer_uri = find_replica(shard_id, info, peer_api_uris, peer_ids)
                resp = requests.post(
                    f"{peer_uri}/collections/{COLLECTION_NAME}/shards/{shard_id}/cleanup?wait=true",
                )
                assert_http_ok(resp)
    
        resp = commit_write_hashring(peer_api_uris[0])
        assert_http_ok(resp)
    
        resp = finish_resharding(peer_api_uris[0])
        assert_http_ok(resp)
    
        wait_for_collection_resharding_operations_count(peer_api_uris[0], COLLECTION_NAME, 0)
    
        # Find peers that still host shards.
        # For "down" with replication_factor=1, the peer that only had shard 2
        # now has no shards and can't serve collection requests.
        # For "up", all peers still have shards (4 shards across 3 peers).
        peers_with_shards = []
        for uri in peer_api_uris:
            cluster_info = get_collection_cluster_info(uri, COLLECTION_NAME)
            if len(cluster_info["local_shards"]) > 0:
                peers_with_shards.append(uri)
    
        # Wait for optimization to complete on peers that have shards
        for uri in peers_with_shards:
>           wait_collection_green(uri, COLLECTION_NAME)

tests/consensus_tests/test_resharding_deferred.py:332: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/consensus_tests/utils.py:608: in wait_collection_green
    raise e
tests/consensus_tests/utils.py:605: in wait_collection_green
    wait_for(check_collection_green, peer_api_uri, collection_name)
tests/consensus_tests/utils.py:738: in wait_for
    while not condition(*args, **kwargs):
              ^^^^^^^^^^^^^^^^^^^^^^^^^^
tests/consensus_tests/utils.py:763: in check_collection_green
    collection_cluster_info = get_collection_info(peer_api_uri, collection_name)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
tests/consensus_tests/utils.py:313: in get_collection_info
    assert_http_ok(r)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

response = <Response [404]>

    def assert_http_ok(response: requests.Response):
        if response.status_code != 200:
            base_msg = f"HTTP request to {response.url} failed with status code {response.status_code} after {response.elapsed.total_seconds()}s"
            if not response.content:
                raise Exception(f"{base_msg} and without response body")
            else:
>               raise Exception(
                    f"{base_msg} with response body:\n{response.json()}")
E               Exception: HTTP request to http://127.0.0.1:37267/collections/test_resharding_deferred failed with status code 404 after 0.015833s with response body:
E               {'status': {'error': 'Not found: status: NotFound, message: "Not found: shard 2 not found", details: [], metadata: MetadataMap { headers: {"content-type": "application/grpc", "date": "Wed, 01 Apr 2026 18:25:10 GMT", "content-length": "0"} } not found'}, 'time': 0.013823993}

tests/consensus_tests/assertions.py:12: Exception

```

--- a/lib/collection/src/shards/shard_holder/mod.rs
+++ b/lib/collection/src/shards/shard_holder/mod.rs
@@ -639,6 +639,20 @@ impl ShardHolder {
                         continue;
                     }
 
+                    // Skip shard being removed by resharding down once the write
+                    // hash ring is committed. The shard is logically gone at this
+                    // point; querying it on a remote peer that already applied
+                    // `finish_resharding` would return a "shard not found" error.
+                    let resharding_removing_down =
+                        self.resharding_state.read().clone().is_some_and(|state| {
+                            state.direction == ReshardingDirection::Down
+                                && state.shard_id == shard_id
+                                && state.stage >= ReshardingStage::WriteHashRingCommitted
+                        });
+                    if resharding_removing_down {
+                        continue;
+                    }
+
                     // Technically, we could skip inactive shards regardless of sharding method,
                     // as we do not expect that shard id can even become inactive on all replicas.
                     // (if it happens, means there is a bug)

--- a/lib/collection/src/shards/shard_holder/mod.rs
+++ b/lib/collection/src/shards/shard_holder/mod.rs
@@ -626,15 +626,16 @@ impl ShardHolder {
                     ShardingMethod::Custom => true,
                 };
 
+                let resharding_state = self.resharding_state.read().clone();
+
                 for (&shard_id, shard) in self.shards.iter() {
                     // Ignore a new resharding shard until it completed point migration
                     // The shard will be marked as active at the end of the migration stage
-                    let resharding_migrating_up =
-                        self.resharding_state.read().clone().is_some_and(|state| {
-                            state.direction == ReshardingDirection::Up
-                                && state.shard_id == shard_id
-                                && state.stage < ReshardingStage::ReadHashRingCommitted
-                        });
+                    let resharding_migrating_up = resharding_state.as_ref().is_some_and(|state| {
+                        state.direction == ReshardingDirection::Up
+                            && state.shard_id == shard_id
+                            && state.stage < ReshardingStage::ReadHashRingCommitted
+                    });
                     if resharding_migrating_up {
                         continue;
                     }
@@ -643,12 +644,11 @@ impl ShardHolder {
                     // hash ring is committed. The shard is logically gone at this
                     // point; querying it on a remote peer that already applied
                     // `finish_resharding` would return a "shard not found" error.
-                    let resharding_removing_down =
-                        self.resharding_state.read().clone().is_some_and(|state| {
-                            state.direction == ReshardingDirection::Down
-                                && state.shard_id == shard_id
-                                && state.stage >= ReshardingStage::WriteHashRingCommitted
-                        });
+                    let resharding_removing_down = resharding_state.as_ref().is_some_and(|state| {
+                        state.direction == ReshardingDirection::Down
+                            && state.shard_id == shard_id
+                            && state.stage >= ReshardingStage::WriteHashRingCommitted
+                    });
                     if resharding_removing_down {
                         continue;
                     }

